MSC2324: Facilitating early releases of software dependent on spec (#2324)
matrix-org/matrix-doc: proposals/2324-when-to-ship.md

     `com.example.new_login`.
   * You can ship the feature at *any* time, so long as you are able to accept
     the technical debt that results from needing to provide adequate backwards
     and forwards compatibility for the vendor prefixed implementation. The
     implementation MUST support the flag disappearing and be generally safe for
     users. Note that implementations early in the MSC review process may also be
     required to provide backwards compatibility with earlier editions of the
     proposal.
   * If you don't want to support the technical debt (or if it's impossible to
     provide adequate backwards/forwards compatibility - e.g. a user authentication
     change which can't be safely rolled back), do not implement the feature and
     wait for Step 7.
   * If at any point the idea changes, the feature flag should also change so
     that implementations can adapt as needed.
3. In parallel, or ahead of implementation, open an MSC and solicit review.
4. Before a FCP (Final Comment Period) can be called, the Spec Core Team will
   require that evidence to prove the MSC works be presented. A typical example

>    of this is an implementation of the MSC (ie: Step 2).

Reviewer: ```suggestion
   of this is an implementation of the MSC (which does not necessarily need to have been shipped anywhere).
```

Step 2 is both how to implement and safely release an MSC into the wild (with compatibility etc). However, a proof-that-MSC-works doesn't have to ship anywhere; it can just be sitting on an unfederated demo HS somewhere or similar to illustrate that it works in practice.

   the specification version, but does have the feature flag, unstable prefixes
   should still be used.
9. A transition period of about 2 months starts immediately after the spec release, before
   implementations start to loudly encourage other implementations to switch to stable
   endpoints. For example, the Synapse team should start asking the Riot team to
   support the stable endpoints (as per Step 8) 2 months after the spec release if they
   haven't already.

It's worth repeating that this process generally only applies if the implementation
wants to ship the feature ahead of the spec being available. By doing so, it takes
on the risk that the spec/MSC may change and it must adapt. If the implementation
is unable to take on that risk, or simply doesn't mind waiting, it should go through
the spec process without shipping an unstable implementation.

To help MSCs get incorporated by implementations as stable features, the spec core
team plans to release the specification more often. How often is undefined and is
largely a case-by-case basis.

To reiterate:

* Implementations MUST NOT use stable endpoints before the MSC is in the spec. This

>   includes NOT using stable endpoints post-FCP.

Reviewer 2: Between fcp and spec release, presumably?

Author: yup

# MSC2324: Facilitating early releases of software dependent on spec

*Note*: This is a process change MSC, not a change to the spec itself.

There's currently an unanswered question by the spec process: when is it
safe to start using stable endpoints or to present a feature as "stable"?
Historically this question would receive very different answers depending
on who you asked, so in an effort to come up with a concise answer the
following process change is proposed.

## Proposal

The new process, from start to finish, is proposed as:

1. Have an idea for a feature.
2. Optionally: implement the feature using unstable endpoints, vendor prefixes,
   and unstable feature flags as appropriate.
   * When using unstable endpoints, they MUST include a vendor prefix. For
     example: `/_matrix/client/unstable/com.example/login`. Vendor prefixes
     throughout this proposal always use the Java package naming convention.

>    * Unstable endpoints **do not** inherit from stable (`/r0`) APIs.

Reviewer 2: I think it might be worth expanding on what this means.

Author: I have added some words

>    * When using an unstable feature flag, they MUST include a vendor prefix.

Reviewer 2: Is it worth clarifying that an unstable feature flag is only required if it is actually useful: for example, a client needing to understand the features provided by a server?

> # MSC2324: Facilitating early releases of software dependent on spec

Author: we should have a worked example/process for how room versions work

Reviewer 3: I don't know if we need to go into too much detail, but it would probably be helpful to have a few examples, e.g.

- changes to room events that affect server operation: need new room version. Use a vendored-prefix room version (e.g. `org.matrix.foo`) when testing.
- changes to adding fields to events that don't affect server operation: Use vendor-prefixed field names when testing (e.g. `"content": {"org.matrix.foo": {...}}`
- new event types that don't affect server operation: Use vendor-prefixed event type when testing
- the stuff from https://github.com/matrix-org/matrix-doc/pull/2324/files#r343621542

Author: it's more the versioning of rooms during development than when to declare a room version. When to declare one is often done at MSC time, and the spec itself is hopefully clear enough for when a new version is needed.

Author: words inc for this^

*Note*: This is a process change MSC, not a change to the spec itself.

There's currently an unanswered question by the spec process: when is it
safe to start using stable endpoints or to present a feature as "stable"?
Historically this question would receive very different answers depending
on who you asked, so in an effort to come up with a concise answer the
following process change is proposed.

## Proposal

The new process, from start to finish, is proposed as:

1. Have an idea for a feature.
2. Optionally: implement the feature using unstable endpoints, vendor prefixes,
   and unstable feature flags as appropriate.
   * When using unstable endpoints, they MUST include a vendor prefix. For
     example: `/_matrix/client/unstable/com.example/login`. Vendor prefixes
     throughout this proposal always use the Java package naming convention.
   * Unstable endpoints **do not** inherit from stable (`/r0`) APIs.
   * When using an unstable feature flag, they MUST include a vendor prefix.

>      This kind of flag shows up in `/versions`. Eg: `com.example.new_login`.

Reviewer 4: The server to server API (nor any other APIs than CS?) doesn't have `/versions`, so right now there's no way to specify feature flags in the server to server API.

Reviewer 2: Indeed. This came up during the team discussion on Tuesday (?). The conclusion was that, if you have a need for a feature flag in those APIs, you'll have to first invent an endpoint that supports them. (The alternative being to handle 404s gracefully and that sort of thing, but michael pointed out that making 404s "normal" is unsatisfactory from the operation POV)

Reviewer 4: @richvdh are you saying "invent an endpoint that supports them" as in, we need to invent a `/versions` endpoint, or we're waiting for something that needs a feature flag to be needed?

Reviewer 2: I don't really understand the question. We'll need to create an endpoint like `/versions` to support feature flags. I don't think there is any intention to do this proactively before we have a feature that needs a feature flag.

Reviewer 4: Right -- are we using a feature flag as an indicator that we support an unstable feature (look before you leap)? If so, the two MSCs regarding data recovery over federation both require them (https://github.com/matrix-org/matrix-doc/pull/2316 to advertise a new namespaced query type and https://github.com/matrix-org/matrix-doc/pull/2314 to advertise an unstable version of the event fetching API).

Author: ftr the rationale for sending @hawkowl here rather than a new MSC is so we can argue about the shape of the s2s feature flags (or lackthereof) somewhere relevant. Given it was left as an unanswered question from when we presented these steps, it seems worthwhile to continue the discussion.

I don't expect it to block this MSC, just to prevent a 5th MSC in the area of "random bits of server information".

Reviewer 2: > Right -- are we using a feature flag as an indicator that we support an unstable feature 

that's pretty much their definition, isn't it? Am I missing something?

> the two MSCs regarding data recovery over federation both require them

I think that's correct, yes

Author: I think I've added some words to deal with this, but am not sure. If I haven't, please open a new thread and we can go from there.

   * You can ship the feature at *any* time, so long as you are able to accept

>      the technical debt. The implementation MUST support the flag disappearing

Reviewer: ```suggestion
     the technical debt that results from needing to provide adequate backwards compatibility for the vendor prefixed implementation. The implementation MUST support the flag disappearing
```

     and be generally safe for users. Note that implementations early in the MSC
     review process may also be required to provide backwards compatibility with
     earlier editions of the proposal.

>    * If you don't/can't support the technical debt, do not implement the feature

Reviewer: ```suggestion
   * If you don't want to support the technical debt (or if it's impossible to provide adequate backwards/forwards compatibility - e.g. a user authentication change which can't be safely rolled back), do not implement the feature
```

     and wait for Step 7.
   * If at any point the idea changes, the feature flag should also change so
     that implementations can adapt as needed.
3. In parallel, or ahead of implementation, open an MSC and solicit review.

> 4. Before an FCP (Final Comment Period) can be called, the Spec Core Team will

Reviewer: ```suggestion
4. Before a FCP (Final Comment Period) can be called, the Spec Core Team will
```

   require that evidence to prove the MSC works be presented. A typical example
   of this is an implementation of the MSC (ie: Step 2).
5. FCP is gone through, and assuming nothing is flagged the MSC lands.
6. A spec PR is written to incorporate the changes into Matrix.
7. A spec release happens.

> 8. Implementations switch to using stable prefixes (ie: `/r0`) if the server

Reviewer: ```suggestion
8. Implementations switch to using stable prefixes (e.g.: `/r0`) if the server
```

   supports the specification version released. If the server doesn't advertise
   the specification version, but does have the feature flag, unstable prefixes
   should still be used.

> 9. A transition period of about 2 months since the spec release happens before

Reviewer: ```suggestion
9. A transition period of about 2 months happens after the spec release, before
```

>    implementations start to encourage other implementations to switch to stable

Reviewer: ```suggestion
   implementations start to loudly encourage other implementations to switch to stable
```

   endpoints. For example, the Synapse team should start asking the Riot team to

>    support the stable endpoints (as per Step 8) 2 months after the spec release.

Reviewer: ```suggestion
   support the stable endpoints (as per Step 8) 2 months after the spec release if they haven't already.
```

It's worth repeating that this process generally only applies if the implementation
wants to ship the feature ahead of the spec being available. By doing so, it takes
on the risk that the spec/MSC may change and it must adapt. If the implementation
is unable to take on that risk, or simply doesn't mind waiting, it should go through

> the normal spec process.

Reviewer: ```suggestion
the spec process without shipping an unstable implementation.
```

To help MSCs get incorporated by implementations as stable features, the spec core
team plans to release the specification more often. How often is undefined and is
largely a case-by-case basis.

To reiterate:

* Implementations MUST NOT use stable endpoints before the MSC is in the spec. This
  includes NOT using stable endpoints post-FCP.
* Implementations CAN ship features that are exposed by default to users before an
  MSC has been merged to the spec, provided they follow the process above.

> * Implementations SHOULD be wary of the technical debt they are incuring by moving

Reviewer: ```suggestion
* Implementations SHOULD be wary of the technical debt they are incurring by moving
```

> # MSC2324: Facilitating early releases of software dependent on spec

Author: This doesn't work for `.well-known` and similar changes.

Reviewer 3: I don't think we should worry too much about edge cases, and just try to apply common sense as much as possible and follow the spirit of this proposal.  There are too many cases that I don't think it's practical to try to figure out something that works for everything.

Author: True, and this isn't to say that the proposal should be blocked on it, just that there's a hole in the process. 

Following the spirit is hard for .well-known because deploying the changes gets difficult, particularly when the changes are behavioural and not structural.

Author: words inc to say "be careful"

*Note*: This is a process change MSC, not a change to the spec itself.

There's currently an unanswered question by the spec process: when is it
safe to start using stable endpoints or to present a feature as "stable"?
Historically this question would receive very different answers depending
on who you asked, so in an effort to come up with a concise answer the
following process change is proposed.

## Proposal

The new process, from start to finish, is proposed as:

1. Have an idea for a feature.

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Reviewer 2: cf https://github.com/matrix-org/matrix-doc/pull/2314#discussion_r340073609

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Reviewer 5: > question: what about changes that add new fields to existing JSON objects (such as API request/response bodies)? Should they be vendor-prefixed?

I would vote yes, as there could be collisions between multiple MSCs in the same way as for APIs. (Though I guess if its behind an unstable flag it matters a bit less)

> likewise changes that make previously-required API parameters optional - should we do anything about that?

Assuming that its a backwards compatible change, I think its probably fine if its behind a flag, though I'm dithering quite a bit.

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Reviewer 2: to be clear, my question was: should an API which returns a backwards-compatible JSON blob be on a `/_matrix/federation/unstable/vendor.prefix.here` path just because it happens to return a couple of extra fields?

requiring that feels like a bunch of unnecessary work, particularly in the case that the calling impl can live without the extra fields.

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Reviewer 2: ftr my opinion on these points is:

* adding new fields to API request or response bodies, or adding new query-params should not require an `unstable` prefix on the API (nor should they require a new `/vN/` prefix once they become canon)
* the new fields in question probably *should* be vendor prefixed until the spec lands
* so should query-params, but I don't know what a vendor-prefixed query-param looks like.

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Reviewer 6: A vendor-prefixed query param is no different from any other vendor-prefixed param, is it?

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Reviewer 2: @KitsuneRal: I guess not.

> 2. Optionally: implement the feature using unstable endpoints, vendor prefixes,

Author: words inc to mirror the 4 thumbs up on vdh's opinion.

   and unstable feature flags as appropriate.
   * When using unstable endpoints, they MUST include a vendor prefix. For
     example: `/_matrix/client/unstable/com.example/login`. Vendor prefixes
     throughout this proposal always use the Java package naming convention.
   * Unstable endpoints **do not** inherit from stable (`/r0`) APIs. Previously,
     one could access the entirety of the Matrix API through `/unstable` however
     this is generally considered a bad practice. Therefore an implementation
     can no longer assume that because its feature-specific endpoint exists that
     any other endpoint will exist in the same unstable namespace.

>    * When using an unstable feature flag, they MUST include a vendor prefix.

Reviewer 3: It seems odd to begin by saying that it MUST include a vendor prefix, and then go on to say that it should only do so under certain conditions.  I think it would be better to just say that it "MUST include a vendor prefix if the client needs to be sure that the server supports the feature."

Author: a variation of this has been applied

>      This kind of flag shows up in `/versions` as, for example, `com.example.new_login`.

Reviewer 3: ```suggestion
     This kind of flag shows up in the `unstable_features` field of `/versions` as, for example, `com.example.new_login`.
```

Author: applied oob
